src/parser.rs

```diff
@@ -333,6 +333,12 @@ pub fn copy_token(ptr: TokenPtr) -> Token {
     ptr.to_owned().into_boxed_slice()
 }
 
+/// Transmutes this token into a Rust string. This is safe for us
+/// because we check that our tokens are valid ASCII (and thus also valid UTF8).
+pub fn as_str(ptr: TokenPtr) -> &str {
+    unsafe { str::from_utf8_unchecked(ptr) }
```

Comment:
TBH I'm not clear on why this is `unsafe` in the first place. Generally type contract violations do not cause memory unsafety, but I guess this one does? Anyway, there *is* a clear abstraction barrier: `Token` and `TokenPtr` are required by contract to store only valid ASCII characters. With this abstraction, `as_str` becomes safe, but `new()` becomes unsafe (or more accurately, coercion from `&[u8]` is unsafe).

This needs a separate PR, but a full encapsulation of `Token` and `TokenPtr`, with no way to create these beyond private methods, should solve the issue.